src/Solvers/multigrid_solver.jl

```diff
@@ -0,0 +1,156 @@
+using Oceananigans.Architectures: architecture
+using Oceananigans.Grids: interior_parent_indices
+using Statistics: norm, dot
+using LinearAlgebra
+using AlgebraicMultigrid: _solve!, init, RugeStubenAMG
+using Oceananigans.Operators: volume, Δyᶠᶜᵃ, Δyᶜᶠᵃ, Δyᶜᶜᵃ, Δxᶠᶜᵃ, Δxᶜᶠᵃ, Δxᶜᶜᵃ, Δyᵃᶜᵃ, Δxᶜᵃᵃ, Δzᵃᵃᶠ, Δzᵃᵃᶜ, ∇²ᶜᶜᶜ
+
+import Oceananigans.Architectures: architecture
+
+mutable struct MultigridSolver{A, G, L, T, M, F}
+               architecture :: A
+                       grid :: G
+                    matrix :: L
+                     abstol :: T
+                     reltol :: T
+                    maxiter :: Int
+              amg_algorithm :: M
+                    x_array :: F
+                    b_array :: F
+end
+
+architecture(solver::MultigridSolver) = solver.architecture
+    
+"""
+    MultigridSolver(linear_operation!::Function,
+                    args...;
+                    template_field::AbstractField,
+                    maxiter = prod(size(template_field)),
+                    reltol = sqrt(eps(eltype(template_field.grid))),
+                    abstol = 0reltol,
+                    amg_algorithm = RugeStubenAMG(),
+                    )
+
+Returns a MultigridSolver that solves the linear equation
+``A x = b`` using a multigrid method, where `A * x` is
+determined by `linear_operation!`
+
+`linear_operation!` is a function with signature `linear_operation!(Ax, x, args...)` 
+that calculates `A * x` for given `x` and stores the result in `Ax`.
+
+
+The solver is used by calling
+
+```
+solve!(x, solver::MultigridSolver, b; kwargs...)
+```
+
+for `solver`, right-hand side `b`, solution `x`, and optional keyword arguments `kwargs...`.
+
+Arguments
+=========
+
+* `template_field`: Dummy field that is the same type and size as `x` and `b`, which
+                    is used to infer the `architecture`, `grid`, and to create work arrays
+                    that are used internally by the solver.
+
+* `maxiter`: Maximum number of iterations the solver may perform before exiting.
+
+* `reltol, abstol`: Relative and absolute tolerance for convergence of the algorithm.
+                    The iteration stops when `norm(A * x - b) < tolerance`.
```

Comment:
Probably we need to update the sentence "The iteration stops when..." to reference `abstol` and `reltol`.

Reply:
Yep, is this better?
```
* `reltol, abstol`: Relative and absolute tolerance for convergence of the algorithm.
                    The iteration stops when `norm(A * x - b) < max(reltol * norm(b), abstol)`.
```

Reply:
It’s great